src/server/gc_worker/gc_worker.rs

```diff
@@ -400,149 +456,224 @@ where
         &mut self,
         keys: Vec<Key>,
         safe_point: TimeStamp,
-        regions_provider: Option<(u64, Arc<dyn RegionInfoProvider>)>,
+        region_or_provider: Either<Region, Arc<dyn RegionInfoProvider>>,
     ) -> Result<(usize, usize)> {
+        let store_id = self.store_id;
         let count = keys.len();
-        let range_start_key = keys.first().unwrap().clone().into_encoded();
+        let range_start_key = keys.first().unwrap().clone();
         let range_end_key = {
             let mut k = keys
                 .last()
                 .unwrap()
                 .to_raw()
                 .map_err(|e| EngineError::Codec(e))?;
             k.push(0);
-            Key::from_raw(&k).into_encoded()
+            Key::from_raw(&k)
         };
 
-        let snapshot = self
-            .engine
-            .snapshot_on_kv_engine(&range_start_key, &range_end_key)?;
-        let mut keys = get_keys_in_regions(keys, regions_provider)?;
-
-        let mut txn = Self::new_txn();
-        let mut reader = if count <= 1 {
-            MvccReader::new(snapshot, None, false)
-        } else {
-            // keys are closing to each other in one batch of gc keys, so do not use
-            // prefix seek here to avoid too many seeks
-            MvccReader::new(snapshot, Some(ScanMode::Forward), false)
+        let (mut handled_keys, mut wasted_keys) = (0, 0);
+        let mut regions = match region_or_provider {
+            Either::Left(region) => vec![region].into_iter().peekable(),
+            Either::Right(region_provider) => get_regions_for_gc(store_id, &keys, region_provider)?
+                .into_iter()
+                .peekable(),
         };
 
-        let (mut handled_keys, mut wasted_keys) = (0, 0);
+        // First item is fetched to initialize the reader and kv_engine
+        let region = regions.peek();
+        if region.is_none() {
+            return Ok((handled_keys, wasted_keys));
+        }
+        let (mut reader, mut kv_engine) = self.create_reader(
+            count,
+            region.unwrap(),
+            range_start_key.clone(),
+            range_end_key.clone(),
+        )?;
+
+        let mut first_iteration = true;
+        let mut txn = Self::new_txn();
         let mut gc_info = GcInfo::default();
-        let mut next_gc_key = keys.next();
-        while let Some(ref key) = next_gc_key {
-            if let Err(e) = self.gc_key(safe_point, key, &mut gc_info, &mut txn, &mut reader) {
-                GC_KEY_FAILURES.inc();
-                error!(?e; "GC meets failure"; "key" => %key,);
-                // Switch to the next key if meets failure.
-                gc_info.is_completed = true;
+        let mut keys = keys.into_iter().peekable();
+        for region in regions {
+            if !first_iteration {
```

Comment:
Why not
```
for region in regions {
    let (reader, kv_engine) = self.create_reader()?;
    let txn = Self::new_txn();
    xxxx;
    Self::flush_txn();
}
```

When one tablet per region, you can't flush transaction across tablet.

Reply:
But it can also have single-rocksdb setting.